### PR TITLE
Deal with OAuth too many redirects

### PR DESCRIPTION
**Pull request by Wauplin**
Fixes: https://github.com/gradio-app/gradio/issues/8533

@abidlabs I tried to have a look at https://github.com/gradio-app/gradio/issues/8533 but I'm not sure how to deal with it. My intuition is that the cookie is forbidden by the browser, ending up in an infinite loop (login => callback => no cookie => login again => callback => no cookie => ...). What we can do is to detect this kind of situation by adding a counter of redirects. Now we have to decide what to do once we know we have this cookie issue. I would be tempted to either:
- return an error to tell the user to configure their browser differently
- redirect to the non-iframe view and hope it will fix the error => **EDIT:** this is the solution we went for

Neither of these are very satisfying solutions in my opinion, even though they are better than ending in a `Too many redirects` error. @abidlabs please let me know what you think.

Note: it's also very difficult to investigate / debug / be sure of what's happening since I've not been able to reproduce this error myself.

### CODE REVIEW

**Comment by abidlabs** [comment]
Did you see this comment in the issue @Wauplin:

> I then tried upgrading the version of gradio to 4.36.1 and starting out at https://matdmiller-hf-oauth.hf.space/ instead of https://huggingface.co/spaces/matdmiller/hf-oauth after the page reloaded and it worked. It seems like something with the default iframe spaces implementation is broken when using huggingface oauth, but not if you visit the second version of the URL.

it sounds like the cookie is not being transferred from the *.hf.space domain to the https://huggingface.co/spaces/* domain. So the issue seems to the Spaces side, but I think the approach of opening the underlying iframe in a new tab might be a good patch for now

File: gradio/oauth.py [hunk 1]
```diff
@@ -105,18 +105,39 @@ async def oauth_redirect_callback(request: fastapi.Request) -> RedirectResponse:
...
+            )
+            if "_target_url" in request.query_params:
+                # Keep same _target_url as before
+                login_uri += "&" + urllib.parse.urlencode({"_target_url": target_url})
+
+            # If the user is redirected more than 3 times, it is very likely that the cookie is not working properly.
+            # (e.g. browser is blocking third-party cookies in iframe). In this case, redirect the user in the
+            # non-iframe view.
+            if nb_redirects > 2:
+                host = os.environ.get("SPACE_HOST")
```
Comment (abidlabs) [inline gradio/oauth.py:131]: Did not know about SPACE_HOST, nice

File: gradio/oauth.py [hunk 2]
```diff
@@ -105,18 +105,39 @@ async def oauth_redirect_callback(request: fastapi.Request) -> RedirectResponse:
...
+
+            # Delete all keys that are related to the OAuth state, just in case
             for key in list(request.session.keys()):
-                # Delete all keys that are related to the OAuth state
                 if key.startswith("_state_huggingface"):
                     request.session.pop(key)
+
+            # Parse query params
+            nb_redirects = int(request.query_params.get("_nb_redirects", 0))
+            target_url = request.query_params["_target_url"]
```
Comment (abidlabs) [inline gradio/oauth.py:116]: Should we `.get()` this as well?

File: gradio/oauth.py [hunk 3]
```diff
@@ -105,18 +105,39 @@ async def oauth_redirect_callback(request: fastapi.Request) -> RedirectResponse:
...
+
+            # Build /login URI with the same query params as before and bump nb_redirects count
+            login_uri = "/login/huggingface"
+            login_uri += "?" + urllib.parse.urlencode(
+                {"_nb_redirects": nb_redirects + 1}
+            )
+            if "_target_url" in request.query_params:
+                # Keep same _target_url as before
+                login_uri += "&" + urllib.parse.urlencode({"_target_url": target_url})
```
Comment (abidlabs) [inline gradio/oauth.py:125]: Could be refactored a bit to be cleaner, e.g.

    query_params = {"_nb_redirects": nb_redirects + 1}
    if target_url:
        query_params["_target_url"] = target_url

    login_uri = f"/login/huggingface?{urlencode(query_params)}"

File: gradio/oauth.py [hunk 4]
```diff
@@ -105,18 +105,39 @@ async def oauth_redirect_callback(request: fastapi.Request) -> RedirectResponse:
...
+
+            # If the user is redirected more than 3 times, it is very likely that the cookie is not working properly.
+            # (e.g. browser is blocking third-party cookies in iframe). In this case, redirect the user in the
+            # non-iframe view.
+            if nb_redirects > 2:
```
Comment (abidlabs) [inline gradio/oauth.py:130]: nit: consider defining a constant MAX_REDIRECTS
```suggestion
            if nb_redirects > MAX_REDIRECTS:
```

**Comment by abidlabs** [comment]
lgtm @Wauplin! just left a few nits above. I'll ping the original issue creator to see if they can test this

File: gradio/oauth.py (hunk 1 above)
Comment (Wauplin) [reply gradio/oauth.py:131]: Here is the reference if you wanna check others :wink: https://huggingface.co/docs/hub/en/spaces-overview#helper-environment-variables

File: gradio/oauth.py (hunk 2 above)
Comment (Wauplin) [reply gradio/oauth.py:116]: Yes :+1:

File: gradio/oauth.py (hunk 3 above)
Comment (Wauplin) [reply gradio/oauth.py:125]: much better indeed! :+1:

File: gradio/oauth.py (hunk 4 above)
Comment (Wauplin) [reply gradio/oauth.py]: addressed in https://github.com/gradio-app/gradio/pull/8951/commits/51b0bac3c33fa9988d81126f2dc1d6a4fe08523b

**Comment by Wauplin** [comment]
Thanks for the review @abidlabs! I've addressed all the points above. Much nicer like this :)

File: gradio/oauth.py (hunk 1 above)
Comment (abidlabs) [reply gradio/oauth.py:131]: Sweet

**Review by abidlabs — APPROVED** [review]
looks good to me! thanks so much @Wauplin, hopefully this fixes it

**Pull request merged**